assets/src/edit-story/components/checklist/checks/videoOptimization.js

```diff
@@ -0,0 +1,218 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { _n, __ } from '@web-stories-wp/i18n';
+import styled from 'styled-components';
+
+/**
+ * Internal dependencies
+ */
+import { useCallback, useMemo, useRef } from 'react';
+import { useFeature } from 'flagged';
+import { useConfig, useLocalMedia, useStory } from '../../../app';
+import { VIDEO_SIZE_THRESHOLD } from '../../../app/media/utils/useFFmpeg';
+import { PRIORITY_COPY } from '../constants';
+import { LayerThumbnail, Thumbnail, THUMBNAIL_TYPES } from '../../thumbnail';
+import {
+  CARD_TYPE,
+  ChecklistCard,
+  DefaultFooterText,
+} from '../../checklistCard';
+import { filterStoryElements } from '../utils/filterStoryElements';
+import {
+  Button,
+  BUTTON_SIZES,
+  BUTTON_TYPES,
+  Tooltip,
+} from '../../../../design-system';
+import { useHighlights } from '../../../app/highlights';
+import { useRegisterCheck } from '../checkCountContext';
+import { StyledVideoOptimizationIcon } from '../../checklistCard/styles';
+
+const STATES = {
+  UPLOADING: 'UPLOADING',
+  UPLOADED: 'UPLOADED',
+};
+
+const OptimizeButton = styled(Button)`
+  margin-top: 4px;
+`;
+
+export function isVideoElementOptimized(element = {}) {
+  if (element.resource?.isTranscoding) {
+    return true;
+  }
+
+  if (element.resource?.local || element.resource?.isOptimized) {
+    return false;
+  }
+
+  const videoArea =
+    (element.resource?.height ?? 0) * (element.resource?.width ?? 0);
+  const isLargeVideo =
+    videoArea >= VIDEO_SIZE_THRESHOLD.WIDTH * VIDEO_SIZE_THRESHOLD.HEIGHT;
+  return isLargeVideo;
+}
+
+export const BulkVideoOptimization = () => {
+  const enableBulkVideoOptimization = useFeature('enableBulkVideoOptimization');
+  const { allowedTranscodableMimeTypes } = useConfig();
+
+  const story = useStory(({ state }) => state);
+  const optimizingResourcesRef = useRef({});
+  const unoptimizedVideos = filterStoryElements(story, (element) => {
+    return (
+      element.type === 'video' &&
+      allowedTranscodableMimeTypes.includes(element.resource?.mimeType) &&
+      isVideoElementOptimized(element)
+    );
+  });
+
+  const setHighlights = useHighlights(({ setHighlights }) => setHighlights);
+
+  const { optimizeVideo } = useLocalMedia((state) => ({
+    optimizeVideo: state.actions.optimizeVideo,
+  }));
+
+  const sequencedVideoOptimization = useCallback(() => {
+    // only attempt to optimize videos that are not currently being transcoded
+    const optimizingResources = unoptimizedVideos.filter(
```

Comment:
`unoptimizedVideos` already filtered out any videos that are being transcoded, no? So this would be redundant

Reply:
I include them in the `unoptimizedVideo` filter because I want them to show while they are loading. Since I want to prevent clicking while the resource is transcoding, I filter them out before triggering the seqeunce.